pep-0701.txt

```diff
@@ -0,0 +1,367 @@
+implementation restructions such as comments, new line characters or
+backslashes. 
+
+The only "surprising" difference is that as the f-string allows to specificy a
+format, expressions that allow a ``:`` character at top level still need to be
+enclosed in parenthesis. This is not new to this work, but is important to
+emphasize that this restriction is still in place. This allows for an easier
+modification of the summary:
+
+    You can place any valid Python expression inside
+    an f-string expression, and everything after a ``:`` character at top level will
+    be identified as a format specification
+
+
+Reference Implementation
+========================
+
+A reference implementation can be found in the implementation_ fork.
+
+Rejected Ideas
+==============
+
+#. We have decided not to lift the restriction that some expression portions
+   need to wrap ``':'`` and ``'!'`` in braces at top level, e.g.::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
+   
+   The reason is that this would this will introduce a considerable amount of
+   complexity for no real benefit. This is due to the fact that the ``:`` character
+   normally separates the f-string format specification. This format specification
+   is currently tokenized as a string. As the tokenizer MUST tokenize what's on the
+   right of the ``:`` as either a string or a stream of tokens this won't allow the
+   parser to differentiate between the different semantics as that would require the
+   tokenizer to backtrack and produce a different set of tokens (this is, first try
+   as a stream of tokens and if it fails try as a string for a format specifier).
+
+   As there is no fundamental advantage in being able to allow lambdas and similar
+   expressions at top level, we have decided to keep the restriction that these must
+   be parenthesized if needed::
+
+    >>> f'Useless use of lambdas: { (lambda x: x*2) }'
+  
+
+Open Issues
+===========
+
+[Any points that are still being decided/discussed.]
+
+
+Footnotes
+=========
+
+
+.. [1] "Grammar for f-strings":
+    https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals
+
+.. [2] "Deferred notice on PEP 536"
+    https://mail.python.org/archives/list/python-dev@python.org/thread/N43O4KNLZW4U7YZC4NVPCETZIVRDUVU2/#NM2A37THVIXXEYR4J5ZPTNLXGGUNFRLZ
+
+.. [3] "Pypy uses now CPython's PEG parser"
+    https://foss.heptapod.net/pypy/pypy/-/commit/fe120f89bf07e64a41de62b224e4a3d80e0fe0d4/pipelines?ref=branch%2Fpy3.9
+
+.. [4] "Python-dev discussion about f-strings in the grammar"
+    https://mail.python.org/archives/list/python-dev@python.org/thread/54N3MOYVBDSJQZTU6MTCPLUPIFSDN5IS/#SAYU6SMP4KT7G7AQ6WVQYUDOSZPKHJMS
+
+.. [5] "Language summit 2022"
+    https://pyfound.blogspot.com/2022/05/the-2022-python-language-summit-f.html
+
+.. [6] Wikipedia article on string interpolation
+    https://en.wikipedia.org/wiki/String_interpolation
```

Comment:
Per #2130 and [PEP 12 guidance](https://peps.python.org/pep-0012/#hyperlinks), actual reST links should be used instead of footnotes for URLs, as it forces the reader through an extra layer of indirection for little benefit.

Also [per PEP 12](https://peps.python.org/pep-0012/#footnotes), if you do use footnotes, named auto-numbered footnotes should always be used instead of hardcoded numbers, as it is easy for the numbers to get out of sync with the footnotes they reference, is tedious to manually renumber them and it ensures the numbers appear in the same order they are first used in the file (currently, they are already well out of order).

Finally, I replaced one hardcoded URL with a direct Intersphinx reference, and the Wikipedia one to link directly to the section being referenced.

```suggestion
.. _official Python grammar: https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals

.. _none of this work was ever implemented: https://mail.python.org/archives/list/python-dev@python.org/thread/N43O4KNLZW4U7YZC4NVPCETZIVRDUVU2/#NM2A37THVIXXEYR4J5ZPTNLXGGUNFRLZ

.. _such as PyPy: https://foss.heptapod.net/pypy/pypy/-/commit/fe120f89bf07e64a41de62b224e4a3d80e0fe0d4/pipelines?ref=branch%2Fpy3.9

.. _discussed on Python-Dev: https://mail.python.org/archives/list/python-dev@python.org/thread/54N3MOYVBDSJQZTU6MTCPLUPIFSDN5IS/#SAYU6SMP4KT7G7AQ6WVQYUDOSZPKHJMS

.. _presented at the Python Language Summit 2022: https://pyfound.blogspot.com/2022/05/the-2022-python-language-summit-f.html

.. _per Wikipedia: https://en.wikipedia.org/wiki/String_interpolation#Examples
```